### PR TITLE
parser: support set operator EXCEPT and INTERSECT

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
refer to [sql-server EXCEPT and INTERSECT](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/set-operators-except-and-intersect-transact-sql?view=sql-server-ver15), we decide to support extra set operators `except` and `intersect`, similar to `union`.

### Check List 

Tests 

 - Unit test
 - Integration test (fail now

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to update the documentation

### CODE REVIEW

**Comment by kennytm** [comment]
(Probably more appropriate to refer to the [MariaDB docs](https://mariadb.com/kb/en/intersect/), which at least share a closer dialect with us)

**Comment by kennytm** [comment]
Both SQL server and MariaDB specify that INTERSECT has higher precedence than UNION and EXCEPT. This PR currently generates a flat list, which is fine, but you need to ensure TiDB recognizes (A &cap; B &cup; C &cap; D) == ((A &cap; B) &cup; (C &cap; D)).

> **Transact-SQL**
> 
> If EXCEPT or INTERSECT is used together with other operators in an expression, it's evaluated in the context of the following precedence:
> 1. Expressions in parentheses
> 2. The INTERSECT operator
> 3. EXCEPT and UNION evaluated from left to right based on their position in the expression



> **MariaDB**
>
> `INTERSECT` has higher precedence than `UNION` and `EXCEPT`. If possible it will be executed linearly but if not it will be translated to a subquery in the `FROM` clause:
>
> ```sql
> (select a,b from t1)
> union
> (select c,d from t2)
> intersect
> (select e,f from t3)
> union
> (select 4,4);
> ```
>
> will be translated to:
>
> ```sql
> (select a,b from t1)
> union
> select c,d from
>   ((select c,d from t2)
>    intersect
>    (select e,f from t3)) dummy_subselect
> union
> (select 4,4)
> ```

**Comment by lzmhhh123** [comment]
> Both SQL server and MariaDB specify that INTERSECT has higher precedence than UNION and EXCEPT. This PR currently generates a flat list, which is fine, but you need to ensure TiDB recognizes (A ∩ B ∪ C ∩ D) == ((A ∩ B) ∪ (C ∩ D)).

Yeah, you are right. I'll make the priority correct in TiDB.

File: ast/dml.go [hunk 1]
```diff
@@ -1067,52 +1073,64 @@ func (n *UnionSelectList) Accept(v Visitor) (Node, bool) {
...
 // See https://dev.mysql.com/doc/refman/5.7/en/union.html
-type UnionStmt struct {
+// See https://mariadb.com/kb/en/intersect/
+// See https://mariadb.com/kb/en/except/
+type SetOprStmt struct {
 	dmlNode
 	resultSetNode
 
-	SelectList *UnionSelectList
+	OptType    SetOprType
```
Comment (kennytm) [inline ast/dml.go:1093]:
```suggestion
	OprType    SetOprType
```
?

IMO it's clearer to spell it out as `SetOperator` than `SetOprType`.

File: parser_test.go [hunk 2]
```diff
@@ -3818,6 +3818,36 @@ func (s *testParserSuite) TestUnion(c *C) {
...
+		{"(select c1 from t1) intersect (select c2 from t2) order by c1 intersect select c3 from t3", false, ""},
+		{"(select c1 from t1) intersect (select c2 from t2) limit 1 intersect select c3 from t3", false, ""},
+		{"(select c1 from t1) intersect select c2 from t2 intersect (select c3 from t3) order by c1 limit 1", true, "(SELECT `c1` FROM `t1`) INTERSECT SELECT `c2` FROM `t2` INTERSECT (SELECT `c3` FROM `t3`) ORDER BY `c1` LIMIT 1"},
+		{"select (select 1 intersect select 1) as a", true, "SELECT (SELECT 1 INTERSECT SELECT 1) AS `a`"},
+		{"select * from (select 1 intersect select 2) as a", true, "SELECT * FROM (SELECT 1 INTERSECT SELECT 2) AS `a`"},
+		{"insert into t select c1 from t1 intersect select c2 from t2", true, "INSERT INTO `t` SELECT `c1` FROM `t1` INTERSECT SELECT `c2` FROM `t2`"},
+		{"insert into t (c) select c1 from t1 intersect select c2 from t2", true, "INSERT INTO `t` (`c`) SELECT `c1` FROM `t1` INTERSECT SELECT `c2` FROM `t2`"},
+		{"select 2 as a from dual intersect select 1 as b from dual order by a", true, "SELECT 2 AS `a` INTERSECT SELECT 1 AS `b` ORDER BY `a`"},
```
Comment (kennytm) [inline parser_test.go:3850]: please add a test case involving a mixture of `intersect`, `except`, `union` and `union all`

Comment (lzmhhh123) [reply parser_test.go:3853]: Done.

File: misc.go [hunk 3]
```diff
@@ -377,6 +377,7 @@ var tokenMap = map[string]int{
 	"INT8":                     int8Type,
 	"INTEGER":                  integerType,
 	"INTERNAL":                 internal,
+	"INTERSECT":                intersect,
```
Comment (fzhedu) [inline misc.go:380]: do `except` need to be inserted here?

File: parser_test.go [hunk 4]
```diff
@@ -3818,6 +3819,43 @@ func (s *testParserSuite) TestUnion(c *C) {
 		{"insert into t select c1 from t1 union select c2 from t2", true, "INSERT INTO `t` SELECT `c1` FROM `t1` UNION SELECT `c2` FROM `t2`"},
 		{"insert into t (c) select c1 from t1 union select c2 from t2", true, "INSERT INTO `t` (`c`) SELECT `c1` FROM `t1` UNION SELECT `c2` FROM `t2`"},
 		{"select 2 as a from dual union select 1 as b from dual order by a", true, "SELECT 2 AS `a` UNION SELECT 1 AS `b` ORDER BY `a`"},
+		// except
+		{"select c1 from t1 except select c2 from t2", true, "SELECT `c1` FROM `t1` EXCEPT SELECT `c2` FROM `t2`"},
```
Comment (fzhedu) [inline parser_test.go:3823]: would `except/union/inserset` be nested in subqueries?

File: parser.y [hunk 5]
```diff
@@ -7825,109 +7827,111 @@ SelectLockOpt:
...
 		}
-		$$ = union
+		$$ = setOpr
 	}
-|	UnionClauseList "UNION" UnionOpt SelectStmtFromTable OrderByOptional SelectStmtLimit SelectLockOpt
+|	SetOprClauseList SetOpr SelectStmtFromTable OrderByOptional SelectStmtLimit SelectLockOpt
 	{
-		st := $4.(*ast.SelectStmt)
-		union := $1.(*ast.UnionStmt)
-		st.IsAfterUnionDistinct = $3.(bool)
```
Comment (fzhedu) [inline parser.y:7873]: can we ignore this?

File: parser_test.go (hunk 4 above)
Comment (lzmhhh123) [reply parser_test.go:3823]: It a type of data source. Could be nested in subqueries.

File: misc.go (hunk 3 above)
Comment (lzmhhh123) [reply misc.go:380]: `except` already exists here.

File: parser.y (hunk 5 above)
Comment (lzmhhh123) [reply parser.y:7873]: I used the `afterSetOprType` instead of it.

**Comment by lzmhhh123** [comment]
@fzhedu PTAL again~

**Pull request merged**